File: frontend/src/engine/parsers/nec-input.ts
/**
 * Build NEC2 card deck from a SimulateAdvancedRequest.
 *
 * TypeScript port of backend/src/simulation/nec_input.py.
 * Used by the WASM engine to generate .nec input locally in the browser.
 */

import type { SimulateAdvancedRequest } from "../types";
import { GROUND_PARAMS } from "../ground";

/**
 * Format a number with Python's :.6g equivalent (6 significant digits,
 * trimming trailing zeros, using exponential notation when needed).
 */
function fmt6g(n: number): string {
  return Number(n.toPrecision(6)).toString();
}

/**
 * Generate a complete NEC2 input card deck from a SimulateAdvancedRequest.
 *
 * Card order: CM, CE, GW, GA, GM, GR, GE, GN, LD, TL, PT, EX, FR, NE, RP, EN
 */
export function buildCardDeck(request: SimulateAdvancedRequest): string {
  const lines: string[] = [];

  // ---- Comment cards ----
  const comment = request.comment ?? "AntennaSim simulation";
  lines.push(`CM ${comment}`);
  lines.push("CE");

  // ---- Geometry Section ----

  // GW cards for each wire
  for (const wire of request.wires) {
    lines.push(
      `GW ${wire.tag} ${wire.segments} ` +
        `${wire.x1.toFixed(6)} ${wire.y1.toFixed(6)} ${wire.z1.toFixed(6)} ` +
        `${wire.x2.toFixed(6)} ${wire.y2.toFixed(6)} ${wire.z2.toFixed(6)} ` +
        `${wire.radius.toFixed(6)}`
    );
  }

  // GA cards for wire arcs
  for (const arc of request.arcs ?? []) {
    lines.push(
      `GA ${arc.tag} ${arc.segments} ` +
        `${arc.arc_radius.toFixed(6)} ${arc.start_angle.toFixed(2)} ${arc.end_angle.toFixed(2)} ` +
        `${arc.wire_radius.toFixed(6)}`
    );
  }

  // GM cards for geometry transforms
  for (const gm of request.transforms ?? []) {
    lines.push(
      `GM ${gm.tag_increment} ${gm.n_new_structures} ` +
        `${(gm.rot_x ?? 0).toFixed(4)} ${(gm.rot_y ?? 0).toFixed(4)} ${(gm.rot_z ?? 0).toFixed(4)} ` +
        `${(gm.trans_x ?? 0).toFixed(6)} ${(gm.trans_y ?? 0).toFixed(6)} ${(gm.trans_z ?? 0).toFixed(6)} ` +
        `${gm.start_tag ?? 0}`
    );
  }

  // GR card for cylindrical symmetry
  if (request.symmetry) {
    lines.push(
      `GR ${request.symmetry.tag_increment} ${request.symmetry.n_copies}`
    );
  }

  // ---- Geometry end ----
  const groundType = request.ground.type;
  if (groundType === "free_space") {
    lines.push("GE -1");
  } else {
    lines.push("GE 0");
  }

  // ---- Program Control Section ----

  // Ground card
  if (groundType === "free_space") {
    lines.push("GN -1");
  } else if (groundType === "perfect") {
    lines.push("GN 1 0 0 0 0 0");
  } else {
    let epsR: number;
    let sigma: number;

    if (groundType === "custom") {
      epsR = request.ground.custom_permittivity ?? 13;
      sigma = request.ground.custom_conductivity ?? 0.005;
    } else {
      const params = GROUND_PARAMS[groundType] ?? GROUND_PARAMS.average!;
      epsR = params.permittivity;
      sigma = params.conductivity;
    }

    lines.push(`GN 2 0 0 0 ${epsR.toFixed(4)} ${sigma.toFixed(6)}`);
  }

  // LD cards (lumped loads)
  for (const ld of request.loads ?? []) {
    lines.push(
      `LD ${ld.load_type} ${ld.wire_tag} ${ld.segment_start} ${ld.segment_end} ` +
        `${fmt6g(ld.param1)} ${fmt6g(ld.param2)} ${fmt6g(ld.param3)}`
    );
  }

  // TL cards (transmission lines)
  for (const tl of request.transmission_lines ?? []) {
    lines.push(
      `TL ${tl.wire_tag1} ${tl.segment1} ${tl.wire_tag2} ${tl.segment2} ` +
        `${tl.impedance.toFixed(4)} ${tl.length.toFixed(6)} ` +
        `${fmt6g(tl.shunt_admittance_real1 ?? 0)} ${fmt6g(tl.shunt_admittance_imag1 ?? 0)} ` +
        `${fmt6g(tl.shunt_admittance_real2 ?? 0)} ${fmt6g(tl.shunt_admittance_imag2 ?? 0)}`
    );
  }

  // PT card (current output control)
  if (request.compute_currents ?? true) {
    lines.push("PT 0 0 0 0"); // Print currents normally
  } else {
    lines.push("PT -1 0 0 0"); // Suppress current printout
  }

  // EX cards (excitations)
  for (const ex of request.excitations) {
    lines.push(
      `EX 0 ${ex.wire_tag} ${ex.segment} 0 ` +
        `${ex.voltage_real.toFixed(4)} ${ex.voltage_imag.toFixed(4)}`
    );
  }

  // ---- Frequency sweep + execution cards ----
  // NEC2 processes cards sequentially: each FR card sets the active frequencies,
  // and the following NE/RP cards trigger computation at those frequencies.
  // For multi-segment sweeps, we emit FR + NE + RP for each segment.

  // Build NE card string (if near-field requested)
  let neCard: string | null = null;
  if (request.near_field) {
    const nf = request.near_field;
    if (nf.plane === "horizontal") {
      const nx = Math.floor(2 * nf.extent_m / nf.resolution_m) + 1;
      const ny = nx;
      const nz = 1;
      neCard =
        `NE 0 ${nx} ${ny} ${nz} ` +
        `${(-nf.extent_m).toFixed(4)} ${(-nf.extent_m).toFixed(4)} ${nf.height_m.toFixed(4)} ` +
        `${nf.resolution_m.toFixed(4)} ${nf.resolution_m.toFixed(4)} ${(0.0).toFixed(4)}`;
    } else {
      const nx = Math.floor(2 * nf.extent_m / nf.resolution_m) + 1;
      const ny = 1;
      const nz = Math.floor(nf.extent_m / nf.resolution_m) + 1;
      neCard =
        `NE 0 ${nx} ${ny} ${nz} ` +
        `${(-nf.extent_m).toFixed(4)} ${(0.0).toFixed(4)} ${(0.0).toFixed(4)} ` +
        `${nf.resolution_m.toFixed(4)} ${(0.0).toFixed(4)} ${nf.resolution_m.toFixed(4)}`;
    }
  }

  // Build RP card string
  const patternStep = request.pattern_step ?? 5;
  const isFreeSpace = groundType === "free_space";
  const thetaStart = isFreeSpace ? -180.0 : -90.0;
  const thetaRange = isFreeSpace ? 360.0 : 180.0;
  const nTheta = Math.floor(thetaRange / patternStep) + 1;
  const nPhi = Math.floor(360 / patternStep);
  const rpCard =
    `RP 0 ${nTheta} ${nPhi} 1000 ` +
    `${thetaStart.toFixed(1)} ${(0.0).toFixed(1)} ` +
    `${patternStep.toFixed(1)} ${patternStep.toFixed(1)}`;

  /** Emit FR + NE + RP cards for one frequency range */
  function emitFrequencyBlock(startMhz: number, stopMhz: number, steps: number): void {
    const stepMhz = steps > 1 ? (stopMhz - startMhz) / (steps - 1) : 0;
    lines.push(`FR 0 ${steps} 0 0 ${startMhz.toFixed(6)} ${stepMhz.toFixed(6)}`);
    if (neCard) lines.push(neCard);
    lines.push(rpCard);
  }

  const freqSegments = request.frequencySegments;
  if (freqSegments && freqSegments.length > 0) {
    const sortedSegments = [...freqSegments].sort((a, b) => a.start_mhz - b.start_mhz);
    for (const seg of sortedSegments) {
      emitFrequencyBlock(seg.start_mhz, seg.stop_mhz, seg.steps);
    }
  } else {
    const freq = request.frequency;
    emitFrequencyBlock(freq.start_mhz, freq.stop_mhz, freq.steps);
  }

  // EN card
  lines.push("EN");

  return lines.join("\n") + "\n";
}

File: backend/src/simulation/nec_input.py
"""Build NEC2 card deck from simulation request models."""

from src.models.simulation import SimulationRequest
from src.models.ground import GroundType


def build_card_deck(request: SimulationRequest) -> str:
    """Generate a complete NEC2 input card deck from a SimulationRequest.

    Supports V1 cards: CM, CE, GW, GE, GN, EX, FR, RP, EN
    Supports V2 cards: LD (loads), TL (transmission lines), PT (current output)

    Returns the full .nec file content as a string.
    """
    lines: list[str] = []

    # Comment cards
    lines.append(f"CM {request.comment}")
    lines.append("CE")

    # ---- Geometry Section ----

    # GW cards for each wire
    for wire in request.wires:
        lines.append(
            f"GW {wire.tag} {wire.segments} "
            f"{wire.x1:.6f} {wire.y1:.6f} {wire.z1:.6f} "
            f"{wire.x2:.6f} {wire.y2:.6f} {wire.z2:.6f} "
            f"{wire.radius:.6f}"
        )

    # GA cards for wire arcs
    for arc in request.arcs:
        lines.append(
            f"GA {arc.tag} {arc.segments} "
            f"{arc.arc_radius:.6f} {arc.start_angle:.2f} {arc.end_angle:.2f} "
            f"{arc.wire_radius:.6f}"
        )

    # GM cards for geometry transforms
    for gm in request.transforms:
        lines.append(
            f"GM {gm.tag_increment} {gm.n_new_structures} "
            f"{gm.rot_x:.4f} {gm.rot_y:.4f} {gm.rot_z:.4f} "
            f"{gm.trans_x:.6f} {gm.trans_y:.6f} {gm.trans_z:.6f} "
            f"{gm.start_tag}"
        )

    # GR card for cylindrical symmetry
    if request.symmetry:
        lines.append(
            f"GR {request.symmetry.tag_increment} {request.symmetry.n_copies}"
        )

    # Geometry end
    # GE 1 if ground-connected vertical (wire touches z=0), GE 0 otherwise
    ground_type = request.ground.ground_type
    if ground_type == GroundType.FREE_SPACE:
        lines.append("GE -1")
    else:
        lines.append("GE 0")

    # ---- Program Control Section ----

    # Ground card
    if ground_type == GroundType.FREE_SPACE:
        lines.append("GN -1")
    elif ground_type == GroundType.PERFECT:
        lines.append("GN 1 0 0 0 0 0")
    else:
        eps_r, sigma = request.ground.get_nec_params()
        lines.append(f"GN 2 0 0 0 {eps_r:.4f} {sigma:.6f}")

    # V2: Loading cards (LD)
    for ld in request.loads:
        # LD TYPE TAG SEG_START SEG_END PARAM1 PARAM2 PARAM3
        lines.append(
            f"LD {ld.load_type.value} {ld.wire_tag} {ld.segment_start} {ld.segment_end} "
            f"{ld.param1:.6g} {ld.param2:.6g} {ld.param3:.6g}"
        )

    # V2: Transmission line cards (TL)
    for tl in request.transmission_lines:
        # TL TAG1 SEG1 TAG2 SEG2 Z0 LENGTH SHUNT_Y1_R SHUNT_Y1_I SHUNT_Y2_R SHUNT_Y2_I
        lines.append(
            f"TL {tl.wire_tag1} {tl.segment1} {tl.wire_tag2} {tl.segment2} "
            f"{tl.impedance:.4f} {tl.length:.6f} "
            f"{tl.shunt_admittance_real1:.6g} {tl.shunt_admittance_imag1:.6g} "
            f"{tl.shunt_admittance_real2:.6g} {tl.shunt_admittance_imag2:.6g}"
        )

    # V2: Current output control
    if request.compute_currents:
        lines.append("PT 0 0 0 0")  # Print currents normally
    else:
        lines.append("PT -1 0 0 0")  # Suppress current printout

    # Excitation cards
    for ex in request.excitations:
        lines.append(
            f"EX 0 {ex.wire_tag} {ex.segment} 0 "
            f"{ex.voltage_real:.4f} {ex.voltage_imag:.4f}"
        )

    # ---- Frequency sweep + execution cards ----
    # NEC2 processes cards sequentially: each FR card sets the active frequencies,
    # and the following NE/RP cards trigger computation at those frequencies.
    # For multi-segment sweeps, we emit FR + NE + RP for each segment.

    # Build NE card string (if near-field requested)
    ne_card: str | None = None
    if request.near_field and request.near_field.enabled:
        nf = request.near_field
        if nf.plane == "horizontal":
            nx = int(2 * nf.extent_m / nf.resolution_m) + 1
            ny = nx
            nz = 1
            x0, y0, z0 = -nf.extent_m, -nf.extent_m, nf.height_m
            dx, dy, dz = nf.resolution_m, nf.resolution_m, 0.0
        else:  # vertical plane along X axis
            nx = int(2 * nf.extent_m / nf.resolution_m) + 1
            ny = 1
            nz = int(nf.extent_m / nf.resolution_m) + 1
            x0, y0, z0 = -nf.extent_m, 0.0, 0.0
            dx, dy, dz = nf.resolution_m, 0.0, nf.resolution_m
        ne_card = f"NE 0 {nx} {ny} {nz} {x0:.4f} {y0:.4f} {z0:.4f} {dx:.4f} {dy:.4f} {dz:.4f}"

    # Build RP card string
    pat = request.pattern
    rp_card = (
        f"RP 0 {pat.n_theta} {pat.n_phi} 1000 "
        f"{pat.theta_start:.1f} {pat.phi_start:.1f} "
        f"{pat.theta_step:.1f} {pat.phi_step:.1f}"
    )

    def emit_frequency_block(start_mhz: float, stop_mhz: float, steps: int) -> None:
        """Emit FR + NE + RP cards for one frequency range."""
        step_mhz = (stop_mhz - start_mhz) / (steps - 1) if steps > 1 else 0.0
        lines.append(f"FR 0 {steps} 0 0 {start_mhz:.6f} {step_mhz:.6f}")
        if ne_card:
            lines.append(ne_card)
        lines.append(rp_card)

    if request.frequency_segments:
        sorted_segments = sorted(request.frequency_segments, key=lambda s: s.start_mhz)
        for seg in sorted_segments:
            emit_frequency_block(seg.start_mhz, seg.stop_mhz, seg.steps)
    else:
        freq = request.frequency
        emit_frequency_block(freq.start_mhz, freq.stop_mhz, freq.steps)

    # End card
    lines.append("EN")

    return "\n".join(lines) + "\n"
